crates/span/src/hygiene.rs

```diff
@@ -21,44 +21,288 @@
+            opaque_and_semitransparent: impl FnOnce(SyntaxContext) -> SyntaxContext,
+        ) -> Self
+        where
+            Db: ?Sized + salsa::Database,
+            Option<MacroCallId>: zalsa_::interned::HashEqLike<T0>,
+            Transparency: zalsa_::interned::HashEqLike<T1>,
+            Edition: zalsa_::interned::HashEqLike<T2>,
+            SyntaxContext: zalsa_::interned::HashEqLike<T3>,
+        {
+            SyntaxContext::ingredient(db).intern(
+                db.as_dyn_database(),
+                StructKey::<'db>(
+                    outer_expn,
+                    outer_transparency,
+                    edition,
+                    parent,
+                    std::marker::PhantomData::default(),
+                ),
+                |id, data| SyntaxContextUnderlyingData {
+                    outer_expn: zalsa_::interned::Lookup::into_owned(data.0),
+                    outer_transparency: zalsa_::interned::Lookup::into_owned(data.1),
+                    edition: zalsa_::interned::Lookup::into_owned(data.2),
+                    parent: zalsa_::interned::Lookup::into_owned(data.3),
+                    opaque: opaque(zalsa_::FromId::from_id(id)),
+                    opaque_and_semitransparent: opaque_and_semitransparent(
+                        zalsa_::FromId::from_id(id),
+                    ),
+                },
+            )
+        }
+
+        /// Invariant: Only [`SyntaxContext::ROOT`] has a [`None`] outer expansion.
+        // FIXME: The None case needs to encode the context crate id. We can encode that as the MSB of
+        // MacroCallId is reserved anyways so we can do bit tagging here just fine.
+        // The bigger issue is that this will cause interning to now create completely separate chains
+        // per crate. Though that is likely not a problem as `MacroCallId`s are already crate calling dependent.
+        pub fn outer_expn<Db>(self, db: &'db Db) -> Option<MacroCallId>
+        where
+            Db: ?Sized + zalsa_::Database,
+        {
+            if self.is_root() {
+                return None;
+            }
+            let fields = SyntaxContext::ingredient(db).fields(db.as_dyn_database(), self);
+            std::clone::Clone::clone(&fields.outer_expn)
+        }
+
+        pub fn outer_transparency<Db>(self, db: &'db Db) -> Transparency
+        where
+            Db: ?Sized + zalsa_::Database,
+        {
+            if self.is_root() {
+                return Transparency::Opaque;
+            }
+            let fields = SyntaxContext::ingredient(db).fields(db.as_dyn_database(), self);
+            std::clone::Clone::clone(&fields.outer_transparency)
+        }
+
+        pub fn edition<Db>(self, db: &'db Db) -> Edition
+        where
+            Db: ?Sized + zalsa_::Database,
+        {
+            // FIXME: is this right?
+            if self.is_root() {
+                return Edition::LATEST;
+            }
```

Comment:
This might be the cause for the syntax highlighting test failures

Reply:
```suggestion
            if self.is_root() {
                return Edition::from_u32(self.0.as_u32() - (SyntaxContext::MAX_ID - Edition::LATEST as u32));
            }
```
Note `Edition::from_u32` isn't a thing, gotta create that as well

Reply:
Good catch! Applying both makes `test_keyword_macro_edition_highlighting` pass, but `test_keyword_highlighting` and `test_rainbow_highlighting` still fail. Not sure why, but working on vendoring the query group macro (and updating it to the latest Salsa...).

Reply:
rainbow i think isnt a problem, we likely are hashing different data now, as long as the hashes are stable and equal as before (wrt to positions they appear in). The other keyword one is still edition related though hmm

Reply:
Actually it is weird that the hashes change, they shouldn't nothing there goes through salsa ...